### ISSUE TITLE
summarysize is incorrect for small union types

### ISSUE THREAD

**Issue by cstjean** [JuliaLang/julia#33675]
```julia
julia> Base.summarysize(Vector{Union{Float64, Missing}}(undef, 1000))
8040

julia> Base.summarysize(Vector{Float64}(undef, 1000))
8040
```
Can that possibly be right? It seems that it's not counting the extra type bit vector?

**Comment by cstjean** [comment]
In fact, for sufficiently large vectors, it's non-deterministic!

```julia
julia> vec = vcat(missing, ones(100000));

julia> Base.summarysize(vec)
1600048

julia> Base.summarysize(vec)
1462016

julia> Base.summarysize(vec)
1600048

julia> Base.summarysize(vec)
1448096

julia> Base.summarysize(vec)
1600048

julia> Base.summarysize(vec)
1458480
```

The docstring says it provides a "lower bound", so technically the output might be correct, but that really looks like a bug somewhere.

**Comment by JeffBezanson** [comment]
The fist issue was fixed by #32886

**Comment by JeffBezanson** [comment]
The changing result is due to GC --- the value is smaller in trials where GC ran. The summarysize code is pulling out each element and boxing it, then using the address to see how many boxes there are. But since those are temporaries, if GC runs they get re-used. But that shouldn't happen since the boxes are unnecessary. I'll fix it.
